### PR TITLE
Hive 25795: Update log4j2 version to 2.16.0 for branch-3 (Naveen Gangam)

### PR DESCRIPTION
**Pull request by nrg4878**
### What changes were proposed in this pull request?
Upgrading some log4j dependencies.

### Why are the changes needed?
Newer releases of log4j have many fixes that are desirable.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT

### CODE REVIEW

File: ql/pom.xml [hunk 1]
```diff
@@ -133,6 +133,11 @@
       <artifactId>log4j-1.2-api</artifactId>
       <version>${log4j2.version}</version>
     </dependency>
+    <dependency>
```
Comment (dengzhhu653) [inline ql/pom.xml:136]: log4j-slf4j-impl has a transitive dependency for log4j-core, may be we do not need this?

**Comment by dengzhhu653** [comment]
+1 for the changes,  maybe we also need to add `-Dlog4j2.formatMsgNoLookups=true` to the services start-scripts.

**Review by marcosValle — CHANGES_REQUESTED** [review]
Version 2.16 has been released: https://lists.apache.org/thread/d6v4r6nosxysyq9rvnr779336yf0woz4

This upgrade fixes a critical security vulnerability, can we prioritize the merge?

File: pom.xml [hunk 2]
```diff
@@ -182,7 +182,7 @@
     <kryo.version>3.0.3</kryo.version>
     <libfb303.version>0.9.3</libfb303.version>
     <libthrift.version>0.9.3</libthrift.version>
-    <log4j2.version>2.10.0</log4j2.version>
+    <log4j2.version>2.15.0</log4j2.version>
```
Comment (marcosValle) [inline pom.xml:185]: 2.16 here

Comment (nrg4878) [reply pom.xml:185]: while it is not required to have 2.16, might as well get it done.

File: ql/pom.xml (hunk 1 above)
Comment (nrg4878) [reply ql/pom.xml]: fixed.

**Review by yongzhi — APPROVED** [review]
Change looks good

**Comment by tasanuma** [comment]
It still seems to depend on log4j 2.6.2 in hive-upgrade-acid module.
```
[INFO] org.apache.hive:hive-upgrade-acid:jar:3.2.0-SNAPSHOT
...
[INFO] +- org.apache.hive:hive-exec:jar:2.3.3:provided
...
[INFO] |  +- org.apache.logging.log4j:log4j-1.2-api:jar:2.6.2:provided
[INFO] |  |  +- org.apache.logging.log4j:log4j-api:jar:2.6.2:provided
[INFO] |  |  \- org.apache.logging.log4j:log4j-core:jar:2.6.2:provided
```

I think it should be excluded in `upgrade-acid/pom.xml`.
```diff
        <dependency>
            <groupId>org.apache.hive</groupId>
            <artifactId>hive-metastore</artifactId>
            <version>2.3.3</version>
            <scope>provided</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.apache.logging.log4j</groupId>
+                    <artifactId>log4j-core</artifactId>
+                </exclusion>
+            </exclusions>
        </dependency>
        <dependency>
            <groupId>org.apache.hive</groupId>
            <artifactId>hive-exec</artifactId>
            <version>2.3.3</version>
            <scope>provided</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.apache.logging.log4j</groupId>
+                    <artifactId>log4j-core</artifactId>
+                </exclusion>
+            </exclusions>
        </dependency>
```

**Comment by tasanuma** [comment]
TestOperationLoggingLayout fails with this change in branch-3.

```
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 25.301 s <<< FAILURE! - in org.apache.hive.service.cli.operation.TestOperationLoggingLayout
[ERROR] testHushableRandomAccessFileAppender(org.apache.hive.service.cli.operation.TestOperationLoggingLayout)  Time elapsed: 6.904 s  <<< ERROR!
java.lang.NoSuchFieldException: appenders

[ERROR] testSwitchLogLayout(org.apache.hive.service.cli.operation.TestOperationLoggingLayout)  Time elapsed: 2.798 s  <<< ERROR!
java.lang.NoSuchFieldException: appenders
```


TestOperationLoggingLayout needs to be fixed in the same way as master branch.
https://github.com/apache/hive/pull/798/files#diff-34f9255a32dec24ce1219b21d225690f0db3f54c013e7c445b37ed8e35036d76

**Review by yongzhi — CHANGES_REQUESTED** [review]
Please fix the issues pointed out by Takanobu Asanuma

**Comment by nrg4878** [comment]
> It still seems to depend on log4j 2.6.2 in hive-upgrade-acid module.
> 
> ```
> [INFO] org.apache.hive:hive-upgrade-acid:jar:3.2.0-SNAPSHOT
> ...
> [INFO] +- org.apache.hive:hive-exec:jar:2.3.3:provided
> ...
> [INFO] |  +- org.apache.logging.log4j:log4j-1.2-api:jar:2.6.2:provided
> [INFO] |  |  +- org.apache.logging.log4j:log4j-api:jar:2.6.2:provided
> [INFO] |  |  \- org.apache.logging.log4j:log4j-core:jar:2.6.2:provided
> ```
> 
> I think it should be excluded in `upgrade-acid/pom.xml`.
> 
> ```diff
>         <dependency>
>             <groupId>org.apache.hive</groupId>
>             <artifactId>hive-metastore</artifactId>
>             <version>2.3.3</version>
>             <scope>provided</scope>
> +            <exclusions>
> +                <exclusion>
> +                    <groupId>org.apache.logging.log4j</groupId>
> +                    <artifactId>log4j-core</artifactId>
> +                </exclusion>
> +            </exclusions>
>         </dependency>
>         <dependency>
>             <groupId>org.apache.hive</groupId>
>             <artifactId>hive-exec</artifactId>
>             <version>2.3.3</version>
>             <scope>provided</scope>
> +            <exclusions>
> +                <exclusion>
> +                    <groupId>org.apache.logging.log4j</groupId>
> +                    <artifactId>log4j-core</artifactId>
> +                </exclusion>
> +            </exclusions>
>         </dependency>
> ```

Thanks @tasanuma for the review. Looks like the scoping for this transitive dependency is "provided" which means it expects the jar to be added by the user. It should not automatically include in the distribution. Is that still a concern?

**Comment by tasanuma** [comment]
@nrg4878 Our security scan software warned us even if it is a provided scope, but as you said, it looks like it doesn't need to be excluded. Thanks.

**Comment by brahmareddybattula** [comment]
@nrg4878 thanks for working on this.. To bump up this we need to have have HIVE-24500 and HIVE-22278....looks as part of the HIVE-22278, following ql/pom.xml change is intentionally missed..?


https://github.com/apache/hive/pull/798/files#diff-652e6b3e0f5adc84cb19dafd6084f4973a41be13e061a9318417046f94462b8eR137

```
<dependency>
      <groupId>org.apache.logging.log4j</groupId>
      <artifactId>log4j-core</artifactId>
      <version>${log4j2.version}</version>
    </dependency> 
```

**Review by yongzhi — APPROVED** [review]
The change looks good. Pending tests

**Comment by tasanuma** [comment]
In the master branch, Hive will update log4j to 2.17.0 by #2894. How about updating it in branch-3, too?

**Comment by tasanuma** [comment]
@nrg4878 Thanks for updating PR. As @brahmareddybattula  said, `ql/pom.xml` depends on log4j 2 by HIVE-22278 in the master branch. I'm not sure if branch-3 needs it or not. Other than that, +1. (non-binding)

File: pom.xml [hunk 3]
```diff
@@ -182,7 +182,7 @@
     <kryo.version>3.0.3</kryo.version>
     <libfb303.version>0.9.3</libfb303.version>
     <libthrift.version>0.9.3</libthrift.version>
-    <log4j2.version>2.10.0</log4j2.version>
+    <log4j2.version>2.17.0</log4j2.version>
```
Comment (achraf-mer) [inline pom.xml:185]:
```suggestion
    <log4j2.version>2.17.1</log4j2.version>
```
To cover CVE-2021-44832

File: standalone-metastore/pom.xml [hunk 4]
```diff
@@ -78,7 +78,7 @@
     <junit.version>4.11</junit.version>
     <libfb303.version>0.9.3</libfb303.version>
     <libthrift.version>0.9.3</libthrift.version>
-    <log4j2.version>2.8.2</log4j2.version>
+    <log4j2.version>2.17.0</log4j2.version>
```
Comment (achraf-mer) [inline standalone-metastore/pom.xml:81]:
```suggestion
    <log4j2.version>2.17.1</log4j2.version>
```
To cover CVE-2021-44832

File: testutils/ptest2/pom.xml [hunk 5]
```diff
@@ -26,7 +26,7 @@ limitations under the License.
   <name>hive-ptest</name>
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-    <log4j2.version>2.10.0</log4j2.version>
+    <log4j2.version>2.17.0</log4j2.version>
```
Comment (achraf-mer) [inline testutils/ptest2/pom.xml:29]:
```suggestion
    <log4j2.version>2.17.1</log4j2.version>
```
To cover CVE-2021-44832

**Review by guptanikhil007 — APPROVED** [review]

**Comment by tasanuma** [comment]
Hi @nrg4878, could you check my previous comment and @achraf-mer 's reviews and update the PR?

**Comment by brahmareddybattula** [comment]
@nrg4878  any update on this..? Can we update to 2.17.1..?

**Pull request closed without merging**